File: daily/2024/1508_Range_Sum_of_Sorted_Subarray_Sums.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    int rangeSum(vector<int>& nums, int n, int left, int right) {
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        for (int i = 0; i < n; i++) pq.push({nums[i], i});
        int round = 1;
        int ans = 0;
        int mod = 1e9+7;
        while(round <= right){
            auto temp = pq.top();
            pq.pop();
            int num = temp.first;
            int index = temp.second;
            if (round >= left) ans = (ans + num) % mod;
            if (index+1 < n) pq.push({num + nums[index+1], index+1});
            round++;
        }
        return ans;
    }
};

int main(){
    vector<int> nums = {1,2,3,4};
    int n = 4;
    int left = 1;
    int right = 5;
    // vector<int> nums = {1,2,3,4};
    // int n = 4;
    // int left = 3;
    // int right = 4;
    // vector<int> nums = {1,2,3,4};
    // int n = 4;
    // int left = 1;
    // int right = 10;
    Solution S;
    cout << S.rangeSum(nums, n, left, right) << endl;
    return 0;
}

File: daily_python/1508_Range_Sum_of_Sorted_Subarray_Sums.py
import sys
import os
from typing import List
import queue

class Solution:
    def rangeSum(self, nums: List[int], n: int, left: int, right: int) -> int:
        pq = queue.PriorityQueue()
        for i in range(n):
            pq.put((nums[i], i))
        ans = 0
        round = 1
        mod = int(1e9+7)
        while round <= right:
            num, index = pq.get()
            if index+1 < n:
                pq.put((num + nums[index+1], index+1))
            if round >= left:
                ans = (ans + num) % mod
            round += 1
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int,lines[0].split('=')[1].strip()[1:-2].split(',')))
            n = int(lines[1].split('=')[1].strip()[0])
            left = int(lines[2].split('=')[1].strip()[0])
            right = int(lines[3].split('=')[1].strip()[0])
            dataset.append((nums, n, left, right))
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        path = os.path.basename(__file__)
        filename = os.path.splitext(path)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums, n, left, right in dataset:
        results.append(solution.rangeSum(nums, n, left, right))
    for index, result in enumerate(results):
        print(f'Example {index+1} : {result}')